third_party/terraform/utils/batcher.go

```diff
@@ -0,0 +1,197 @@
+// implementations per type of request per service.
+// Function type for combine existing batches and additional batch data
+type batcherCombineFunc func(body interface{}, toAdd interface{}) (interface{}, error)
+
+// Function type for sending a batch request
+type batcherSendFunc func(resourceName string, body interface{}) (interface{}, error)
+
+// batchResponse bundles an API response (data, error) tuple.
+type batchResponse struct {
+	body interface{}
+	err  error
+}
+
+// startedBatch refers to a processed batch whose timer to send the request has
+// already been started. The responses for the request is sent to each listener
+// channel, representing parallel callers that are waiting on requests
+// combined into this batch.
+type startedBatch struct {
+	*BatchRequest
+
+	listeners []chan batchResponse
+	timer     *time.Timer
+}
+
+// batchingConfig contains user configuration for controlling batch requests.
+type batchingConfig struct {
+	sendAfter       time.Duration
+	disableBatching bool
+}
+
+// Initializes a new batcher
+func NewRequestBatcher(config *batchingConfig) *RequestBatcher {
+	return &RequestBatcher{
+		batchingConfig: config,
+		batches:        make(map[string]*startedBatch),
+	}
+}
+
+// SendRequestWithTimeout is expected to be called per parallel call.
+// It manages waiting on the result of a batch request.
+func (b *RequestBatcher) SendRequestWithTimeout(batchType string, request *BatchRequest, timeout time.Duration) (interface{}, error) {
+	if request == nil {
+		return nil, fmt.Errorf("error, cannot request batching for nil BatchRequest")
+	}
+	if request.CombineF == nil {
+		return nil, fmt.Errorf("error, cannot request batching for BatchRequest with nil CombineF")
+	}
+	if request.SendF == nil {
+		return nil, fmt.Errorf("error, cannot request batching for BatchRequest with nil SendF")
+	}
+	if b.disableBatching {
+		log.Printf("[DEBUG] Batching is disabled, sending single request for %q", request.DebugId)
+		return request.SendF(request.ResourceName, request.Body)
+	}
+
+	respCh, err := b.startBatchRequest(batchType, request)
+	if err != nil {
+		return nil, fmt.Errorf("error adding request to batch: %s", err)
+	}
+
+	timer := time.NewTimer(timeout)
+
+	select {
+	case resp := <-respCh:
+		if resp.err != nil {
+			return nil, fmt.Errorf("Batch %q Request %q returned error: %v", batchType, request.DebugId, resp.err)
+		}
+		return resp.body, nil
+	case <-timer.C:
+		break
+	}
+	return nil, fmt.Errorf("Request %s timed out after %v", batchType, timeout)
+}
+
+// startBatchRequest manages batching logic that access shared information
+// (i.e. existing batches)
+func (b *RequestBatcher) startBatchRequest(batchType string, newRequest *BatchRequest) (<-chan batchResponse, error) {
+	b.Lock()
+	defer b.Unlock()
+
+	// The calling goroutine will need a channel to wait on for a response.
+	respCh := make(chan batchResponse, 1)
+
+	batchId := fmt.Sprintf("%s:%s", batchType, newRequest.ResourceName)
```

Comment:
Can you add a comment (probably to the header for `SendRequestWithTimeout`) explaining how batchable requests are grouped? It's available here, but that information would also be useful on the user-facing fn.

Reply:
Do you think it would be a good idea to move/copy resource name out of request so you give (batchType, resourceName, request, timeout)? I think it might be more clear that way.

Reply:
I'll just make batchType batchKey, and the caller will control how to key the batch

Reply:
I'd just like to see a summary of (typical) rules of batching before digging into the function so that when developers are looking @ resources with batched requests it's obvious how they're batched!

Pulling the name from the request is sensible, I just want to know that'll happen without needing to click into a few different functions.